mlfinlab/microstructural_features/feature_generator.py

```diff
@@ -0,0 +1,301 @@
+        :param batch_size: (int) Number of rows to read in from the csv, per batch.
+        :param volume_encoding: (dict) Dictionary of encoding trades size anc calculate entropy on encoded messages
+        :param pct_encoding: (dict) Dictionary of encoding log returns anc calculate entropy on encoded messages
+        """
+
+        if isinstance(trades_input, str):
+            self.generator_object = pd.read_csv(trades_input, chunksize=batch_size, parse_dates=[0])
+            # Read in the first row & assert format
+            first_row = pd.read_csv(trades_input, nrows=1)
+            self._assert_csv(first_row)
+        elif isinstance(trades_input, pd.DataFrame):
+            self.generator_object = _crop_data_frame_in_batches(trades_input, batch_size)
+        else:
+            raise ValueError('trades_input is neither string(path to a csv file) nor pd.DataFrame')
+
+        # Base properties
+        self.bar_index_iterator = iter(bar_index)
+        self.current_date_time = self.bar_index_iterator.__next__()
+
+        # Cache properties
+        self.price_diff = []
+        self.trade_size = []
+        self.tick_rule = []
+        self.dollar_size = []
+        self.log_ret = []
+
+        # Entropy properties
+        self.volume_encoding = volume_encoding
+        self.pct_encoding = pct_encoding
+
+        # Batch_run properties
+        self.prev_price = None
+        self.prev_tick_rule = 0
+
+    def get_features(self, verbose=True, to_csv=False, output_path=None):
+        """
+        Reads a csv file of ticks or pd.DataFrame in batches and then constructs corresponding microstructural intra-bar features:
+        average tick size, tick rule sum, VWAP, Kyle lambda, Amihud lambda, Hasbrouck lambda, tick/volume/pct Shannon, Lempel-Ziv,
+        Plug-in entropies if corresponding mapping dictionaries are provided (self.volume_encoding, self.pct_encoding).
+        The csv file must have only 3 columns: date_time, price, & volume.
+
+        :param verbose: (Boolean) Flag whether to print message on each processed batch or not
+        :param to_csv: (Boolean) Flag for writing the results of bars generation to local csv file, or to in-memory DataFrame
+        :param output_path: (Boolean) Path to results file, if to_csv = True
+        :return: (DataFrame or None) of microstructural features for bar index
+        """
+
+        if to_csv is True:
+            header = True  # if to_csv is True, header should be written on the first batch only
+            open(output_path, 'w').close()  # Clean output csv file
+
+        # Read csv in batches
+        count = 0
+        final_bars = []
+        cols = ['date_time', 'avg_tick_size', 'tick_rule_sum', 'vwap', 'kyle_lambda', 'amihud_lambda',
+                'hasbrouck_lambda']
+
+        # Entropy features columns
+        for en_type in ['shannon', 'plug_in', 'lempel_ziv']:
```

Comment:
Awaiting for the future enthusiasts that might create new entropies I would suggest to define this list one time in __init__ to ease code update.